openstack/baremetal/v1/nodes/requests.go

```diff
@@ -371,3 +387,52 @@ func GetSupportedBootDevices(client *gophercloud.ServiceClient, id string) (r Su
 	})
 	return
 }
+
+// A cleaning step has required keys ‘interface’ and ‘step’, and optional key ‘args’. If specified,
+// the value for ‘args’ is a keyword variable argument dictionary that is passed to the cleaning step
+// method.
+type CleanStep struct {
+	Interface string            `json:"interface,required"`
+	Step      string            `json:"step,required"`
+	Args      map[string]string `json:"args,omitempty"`
+}
+
+// ProvisionStateOptsBuilder allows extensions to add additional parameters to the
+// ChangeProvisionState request.
+type ProvisionStateOptsBuilder interface {
+	ToProvisionStateMap() (map[string]interface{}, error)
+}
+
+// ProvisionStateOpts for a request to change a node's provision state. A config drive should be base64-encoded
+// gzipped ISO9660 image.
+type ProvisionStateOpts struct {
+	Target         TargetProvisionState `json:"target,required"`
+	ConfigDrive    string               `json:"configdrive,omitempty"`
+	CleanSteps     []CleanStep          `json:"clean_steps,omitempty"`
+	RescuePassword string               `json:"rescue_password,omitempty"`
+}
+
+// ToProvisionStateMap assembles a request body based on the contents of a CreateOpts.
+func (opts ProvisionStateOpts) ToProvisionStateMap() (map[string]interface{}, error) {
```

Comment:
One thing I worry about here is that `ConfigDrive` could potentially be up to 64MB.  Should this be a pointer receiver? In practice it'll be far smaller (a few k.b. at maximum)

Reply:
64mb is large, but unfortunately I don't know if it's large enough to cause issues.

The alternative to using a string is to make this an `io.Reader` like it's done for object creation: https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L148

The Reader then gets streamed into the request body here: https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L240

This allows data to be read from disk and streamed to the API at the same time, thus reducing memory consumption.

Reply:
I was thinking more of golang using pass-by-value and having to pass around the string reference of the configdrive value, instead of a pointer to the string.  But, according to https://github.com/golang/go/wiki/CodeReviewComments#receiver-type, using a value receiver is fine.

In any case, I would guess it's probably fine, no one is really sending 64mb configdrives in the real world for the vast majority of cases, it's all text data for cloud-init or iginition.